File: src/main/java/com/scottscmo/ppt/BibleTemplateHandler.java
package com.scottscmo.ppt;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import com.scottscmo.model.bible.BibleModel;
import com.scottscmo.model.bible.BibleReference;
import com.scottscmo.model.bible.BibleVerse;

import org.apache.poi.xslf.usermodel.XMLSlideShow;
import org.apache.poi.xslf.usermodel.XSLFSlide;
import org.apache.poi.xslf.usermodel.XSLFSlideMaster;

public class BibleTemplateHandler implements TemplateHandler {
    private BibleModel bibleModel = BibleModel.getInstance();

    /**
     * expression format: e.g. {bible} cuv,niv - john 1:2-3;2:1-2
     */
    @Override
    public void evaluateTemplate(XMLSlideShow slides, int index) {
        XSLFSlide srcSlide = slides.getSlides().get(index);
        String bibleReference = findText(srcSlide, "{bible}");
        if (bibleReference != null) {
            insertBibleText(slides, index, bibleReference.substring(7).trim());
            System.out.println("Inserting bible text at " + srcSlide.getSlideNumber());
        }
    }

    public void insertBibleText(XMLSlideShow slides, int index, String bibleReferenceStr) {
        if (slides == null || bibleReferenceStr == null) return;
        insertBibleText(slides, index, new BibleReference(bibleReferenceStr));
    }

    public void insertBibleText(XMLSlideShow slides, int index, BibleReference ref) {
        if (slides == null || ref == null) return;

        Map<String, List<BibleVerse>> bibleVerses = bibleModel.getBibleVerses(ref);

        if (bibleVerses == null) return;

        XSLFSlideMaster masterSlide = slides.getSlideMasters().get(0);

        Map<String, String> bookNames = bibleModel.getBookNames(ref.getBook());

        // create title slide
        XSLFSlide titleSlide = slides.createSlide(masterSlide.getLayout("title"));
        for (String version : ref.getVersions()) {
            replaceText(titleSlide, "{title_" + version + "}", bookNames.get(version));
        }
        replaceText(titleSlide, "{range}", ref.getRanges().stream().map(range -> range.toString()).collect(Collectors.joining(";")));

        // create verse slides
        int numVerses = bibleVerses.get(ref.getVersions()[0]).size();
        for (int i = 0; i < numVerses; i++) {
            for (String version : ref.getVersions()) {
                XSLFSlide slide = slides.createSlide(masterSlide.getLayout("verse_" + version));
                BibleVerse verse = bibleVerses.get(version).get(i);
                String refStr = String.format("%s %d:%d", bookNames.get(version), verse.chapter(), verse.verse());

                replaceText(slide, "{verse}", verse.verse() + " " + verse.text());
                replaceText(slide, "{title}", refStr);
            }
        }
    }
}

File: src/main/kotlin/com/scottscmo/ppt/BibleTemplateHandler.kt
package com.scottscmo.ppt

import com.scottscmo.model.bible.BibleModel
import com.scottscmo.model.bible.BibleReference
import org.apache.poi.xslf.usermodel.XMLSlideShow
import org.apache.poi.xslf.usermodel.XSLFSlideLayout

class BibleTemplateHandler {
    private val bibleModel: BibleModel = BibleModel.get()
    private val titleMasterKey = "title"
    private val verseMasterKeyPrefix = "verse"
    private val mainLayoutKey = "main"

    fun insertBibleText(ppt: XMLSlideShow, bibleReferenceStr: String) {
        insertBibleText(ppt, BibleReference(bibleReferenceStr))
    }

    fun insertBibleText(ppt: XMLSlideShow, ref: BibleReference) {
        val bibleVerses = bibleModel.getBibleVerses(ref)
        val bookNames = bibleModel.getBookNames(ref.book)
        requireNotNull(bookNames) { "Unable to query book names with book ${ref.book}" }

        // create title slide
        val titleLayout = TemplatingUtil.getSlideMasterLayout(ppt, titleMasterKey, mainLayoutKey)
        requireNotNull(titleLayout) { "Missing title master slide" }

        val titleSlide = ppt.createSlide(titleLayout)
        val titleSlideValues: MutableMap<String, String> = mutableMapOf()
        for (version in ref.versions) {
            titleSlideValues["{title_$version}"] = bookNames.getOrDefault(version, "")
        }
        titleSlideValues["{range}"] = ref.ranges.joinToString(";")
        TemplatingUtil.replacePlaceholders(titleSlide, titleSlideValues)

        // create verse slides
        val verseLayouts: MutableMap<String, XSLFSlideLayout?> = mutableMapOf()
        for (version in ref.versions) {
            val key = "${this.verseMasterKeyPrefix}_$version"
            verseLayouts[key] = TemplatingUtil.getSlideMasterLayout(ppt, key, mainLayoutKey)
        }

        val numVerses = bibleVerses[ref.versions[0]]?.size ?: 0
        for (i in 0 until numVerses) {
            for (version in ref.versions) {
                val slide = ppt.createSlide(verseLayouts["${this.verseMasterKeyPrefix}_$version"])
                val verse = bibleVerses[version]!![i]
                val refStr = String.format("%s %d:%d", bookNames[version], verse.chapter, verse.index)
                TemplatingUtil.replacePlaceholders(slide, mapOf(
                    "{verse}" to verse.index.toString() + " " + verse.text,
                    "{title}" to refStr
                ))
            }
        }
    }
}